### PR TITLE
helm: Add check for prometheus service monitoring CRDs

### PR DESCRIPTION
**Pull request by sayboras**
## Description 

helm: Add check for prometheus service monitoring CRDs

Service Monitnor CRD (e.g. monitoring.coreos.com/v1) is no longer
managed and installed as part of cilium chart. This commit is to
add validation to check if monitoring.coreos.com/v1 CRD is required
and available before installation.

Relates: #13513

**TODO**: Update docs with step by step installation, maybe on this page https://docs.cilium.io/en/v1.9.0-rc2/operations/metrics/#installation, I will do it in separate PR. 

```release-note
helm: Add check for prometheus service monitoring CRDs
```

### CODE REVIEW

**Review by aanm — CHANGES_REQUESTED** [review]
This change is good but I don't think this is enough.

@seanmwinn what was the rational behind removing the service monitor CRD from the Cilium charts? As far as I remember users had issues when deploying Cilium on a cluster with this CRD. More info https://github.com/cilium/cilium/pull/11261

**Comment by sayboras** [comment]
@aanm few points hightlighted before in helm refractor PR https://github.com/cilium/cilium/pull/13259#issuecomment-707041569.

File: install/kubernetes/cilium/values.yaml [hunk 1]
```diff
@@ -446,6 +446,9 @@ hubble:
     port: 9091
     # Creates ServiceMonitor resources for Prometheus Operator
     serviceMonitor:
+      ## Enable service monitor. Required prometheus CRS available.
```
Comment (diversario) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
      ## Enable service monitor. Required prometheus CRD available.
```
?

**Comment by diversario** [comment]
> This change is good but I don't think this is enough.
> 
> @seanmwinn what was the rational behind removing the service monitor CRD from the Cilium charts? As far as I remember users had issues when deploying Cilium on a cluster with this CRD. More info #11261
 
I see that the Prometheus ServiceMonitor CRD was removed already. It makes sense that Cilium shouldn't be owning someone else's CRDs, but it does make initial provisioning of clusters difficult – I think basically it requires a second provisioning pass to enable servicemonitors for Cilium. I don't know the best way to address this. Almost makes me think a separate chart just for servicemonitors would be ok, so that cluster provisioning would go `cilium` -> `prometheus` -> `cilium-monitoring`.

**Comment by sayboras** [comment]
> I see that the Prometheus ServiceMonitor CRD was removed already. It makes sense that Cilium shouldn't be owning someone else's CRDs, but it does make initial provisioning of clusters difficult – I think basically it requires a second provisioning pass to enable servicemonitors for Cilium. I don't know the best way to address this. Almost makes me think a separate chart just for servicemonitors would be ok, so that cluster provisioning would go `cilium` -> `prometheus` -> `cilium-monitoring`.

It's a very fair point :thinking:, having a sequence of installation will help, but really not a good user experience I can say. Let me do some research on this, while waiting for others to bouncing ideas here.

**Comment by aanm** [comment]
@sayboras, me and @diversario spent quite some time trying to figure out a good solution for this problem so we end up adding the CRD into the Cilium chart since from the helm documentation the CRDs are only installed if they don't exist and they are never removed. It's really tricky to solve this problem in a different way but let us know about your findings!

**Comment by diversario** [comment]
> > I see that the Prometheus ServiceMonitor CRD was removed already. It makes sense that Cilium shouldn't be owning someone else's CRDs, but it does make initial provisioning of clusters difficult – I think basically it requires a second provisioning pass to enable servicemonitors for Cilium. I don't know the best way to address this. Almost makes me think a separate chart just for servicemonitors would be ok, so that cluster provisioning would go `cilium` -> `prometheus` -> `cilium-monitoring`.
> 
> It's a very fair point 🤔, having a sequence of installation will help, but really not a good user experience I can say. Let me do some research on this, while waiting for others to bouncing ideas here.

I'm probably way overthinking this but... what if Cilium Operator was in charge of creating the ServiceMonitor. As in:

1. I deploy Cilium with `serviceMonitor.enabled=true`
1. Cilium Operator configuration records that user asked to create a SM
1. Operator checks if it is possible (if CRD is available), and if yes – creates it; if not – logs a warning, waits some time, rechecks. Repeat forever until SM is created.

The nice part of deploying the SM as part of the Helm chart is that it's very easy to modify it and it doesn't depend on the Cilium image being used (it was very easy for me to modify it in my local fork of the Helm chart). But as far as user experience goes, this might be an edge case and making it possible to just install Cilium in one pass is more important.

**Comment by sayboras** [comment]
> I'm probably way overthinking this but... what if Cilium Operator was in charge of creating the ServiceMonitor ...

Feel like the Service Monitor CRDs still needs to be created by users manually if I am not wrong.

As we don't need to have cilium ready _before_ installing ServiceMonitoring CRDs, can we perform simple check in validate.yaml ?

- If `monitoring.coreos.com/v1` is required but not available, fail `helm install` with some info/steps (e.g. kubectl apply -f https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml).
- If `monitoring.coreos.com/v1` is not required (e.g. monitor related flags `enabled` = false), just carry on as usual.

Basically, I want to achieve below points, feel free to let me know if my assumption is wrong.

- Remove Service Monitoring in cilium helm chart. The main reason is about functional ownership.
- Keep user experience fairly with minimun changes.

**Comment by diversario** [comment]
> > I'm probably way overthinking this but... what if Cilium Operator was in charge of creating the ServiceMonitor ...
> 
> Feel like the Service Monitor CRDs still needs to be created by users manually if I am not wrong.

Yes, and it is expected; SM CRD is installed as part of Prometheus Operator, so in the course of users setting up their monitoring stack.

> As we don't need to have cilium ready _before_ installing ServiceMonitoring CRDs, can we perform simple check in validate.yaml ?
> 
> * If `monitoring.coreos.com/v1` is required but not available, fail `helm install` with some info/steps (e.g. SM CRDs not available, please install it by your preferred method or run kubectl apply -f [prometheus-operator/prometheus-operator@`master`/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml (raw)](https://raw.githubusercontent.com/prometheus-operator/prometheus-operator/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml)).
> * If `monitoring.coreos.com/v1` is not required (e.g. monitor related flags `enabled` = false), just carry on as usual.
> 
> Basically, I want to achieve below points, feel free to let me know if my assumption is wrong.
> 
> * Remove Service Monitoring in cilium helm chart. The main reason is about functional ownership.
> * Keep user experience fairly with minimal changes.

For the scope of this PR that makes sense. The functionality I described in the previous comment is definitely way outside of the scope of this PR.

It will be a breaking change as far as Helm chart goes, though, because if anyone expects the chart to get installed with the monitoring enabled will see it fail. Not the end of the world, but an annoyance at least.

**Review by aanm — APPROVED** [review]

**Comment by joestringer** [comment]
One point on this would be, how do we expect the overall flow for install to work?

Option A:
* Install monitoring
* Install Cilium

Option B:
* Install Cilium without monitoring
* Install monitoring
* Modify Cilium to enable monitoring

And how do we tell users to set up Cilium with monitoring in the documentation?

In some ways it seems nice to figure out during install whether the prerequisites are satisfied and let the user know since it makes the requirements more visible at install time. Sometimes though these kinds of checks can cause more trouble down the road because it forces a certain ordering of operations.

Is it possible for us to provide a runtime warning or something instead to signal when the user has installed Cilium but hasn't set up the monitoring CRDs? Then ideally whenever the user installs the CRD, Cilium would automatically just start working with it (but I could understand if a first cut also requires restarting Cilium).

**Comment by aanm** [comment]
@joestringer the problem was, as far as I remember, the monitoring framework depends on Cilium to start.

**Comment by joestringer** [comment]
@aanm and what's the behaviour from Cilium side right now if the monitor CRD is not available but Cilium is configured with monitoring?

**Comment by aanm** [comment]
> @aanm @sayboras and what's the behaviour from Cilium side right now if the monitor CRD is not available but Cilium is configured with monitoring?

In 1.8 Cilium deploys the missing CRD, removed by the helm refactoring, in 1.9 / master Cilium won't install it because it does not exist.

**Comment by joestringer** [comment]
@aanm what's the consequences if the CRD is missing? Cilium presumably runs fine and does its core functions (ie, handle networking & policy), I assume that monitoring just doesn't work. If someone upgrades, the CRD is already in place so I guess existing users are fine? Then it's only new users which need to take additonal action, so we can document how to do it right for new installs.

Then beyond that I'd guess there's the corner cases if someone gets into this state and doesn't know how to debug it, how do we make it obvious to users that the CRD is missing and they need to take action? This PR presents one approach - force it during install - but there could be other approaches, like just documenting it as a troubleshooting step, or trying to perform runtime detection of the missing CRD when Cilium is configured with monitoring enabled, then logging a warning in that case.

File: install/kubernetes/cilium/templates/validate.yaml [hunk 2]
```diff
@@ -9,3 +9,10 @@
     {{ fail "Hubble Relay requires .Values.hubble.enabled=true" }}
   {{- end }}
 {{- end }}
+
+{{/* validate service monitoring CRDs */}}
+{{- if not (.Capabilities.APIVersions.Has "monitoring.coreos.com/v1") }}
+  {{- if and (.Values.prometheus.enabled) (or (.Values.prometheus.serviceMonitor.enabled) (.Values.operator.prometheus.serviceMonitor.enabled)) }}
```
Comment (kkourt) [inline install/kubernetes/cilium/templates/validate.yaml]: nit: This feels more readable to me if we reverse the two checks. I.e., only check APIVersions for "monitoring.coreos.com/v1" if `serviceMonitor` is enabled. What do you think?

Also, for my understanding: what is the reason that we need to check for `.Values.prometheus.enabled`? Is it because if this is not enabled, then accessing `.Values.prometheus.serviceMonitor.enabled` will cause an error?

File: install/kubernetes/cilium/values.yaml [hunk 3]
```diff
@@ -452,6 +452,9 @@ hubble:
     port: 9091
     # Creates ServiceMonitor resources for Prometheus Operator
     serviceMonitor:
+      ## Enable service monitor. Required prometheus CRDs available.
```
Comment (kkourt) [inline install/kubernetes/cilium/values.yaml]: How about: 
```
# Enable service monitors.
# This requires the prometheus CRDs to be available (see https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml)

**Review by kkourt — COMMENTED** [review]
Some minor comments from my side.

File: install/kubernetes/cilium/templates/validate.yaml [hunk 4]
```diff
@@ -9,3 +9,10 @@
...
+      {{ fail "Service Monitor requires monitoring.coreos.com/v1 CRDs. Please refer to https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml" }}
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/validate.yaml:16]: Is pointing towards a master branch the best option here or should we just pick the most recent release and then only bump if we need newer features? Just wondering if it's possible that in future prometheus either moves this example path or replaces it with a v2 CRD or something.

Not such a big deal though.

**Review by joestringer — APPROVED** [review]
@sayboras pointed out out-of-band that without this check, `helm install` will just fail with a more cryptic error. Short of reinstating the monitoring CRD back into our charts, I don't see a great alternative.

**Comment by sayboras** [comment]
Thanks guys for your inputs and comments, as now we are kind of agree on the direction, I will close the loop and wrap it up for this PR :100:

File: install/kubernetes/cilium/templates/validate.yaml (hunk 2 above)
Comment (sayboras) [reply install/kubernetes/cilium/templates/validate.yaml]: > This feels more readable to me if we reverse the two checks. I.e., only check APIVersions for "monitoring.coreos.com/v1" if serviceMonitor is enabled. What do you think?

Good point, it makes sense to reverse two checks (e.g. static first, then runtime later). I have incorporated this point :100: 

> Also, for my understanding: what is the reason that we need to check for .Values.prometheus.enabled? Is it because if this is not enabled, then accessing .Values.prometheus.serviceMonitor.enabled will cause an error?

ServiceMonitoring CRDs is only created if .Values.prometheus.enabled is true right now, hence I just want to keep the same check.

Comment (kkourt) [reply install/kubernetes/cilium/templates/validate.yaml]: > ServiceMonitoring CRDs is only created if .Values.prometheus.enabled is true right now, hence I just want to keep the same check.

Makes perfect sense! Thanks!

**Review by kkourt — APPROVED** [review]

**Comment by diversario** [comment]
> @aanm what's the consequences if the CRD is missing? Cilium presumably runs fine and does its core functions (ie, handle networking & policy), I assume that monitoring just doesn't work. If someone upgrades, the CRD is already in place so I guess existing users are fine? Then it's only new users which need to take additonal action, so we can document how to do it right for new installs.

That's all true, and doable. The downside of this is that it breaks automated cluster provisioning flows if there's a requirement to go back to the provisioning config, update it, and run it again as a second pass. And new cluster provisioning can be a very common thing when it's done for CI, one-off testing or anything like that. So ideally, it'd be great to have a single-pass provisioning.
